### PR TITLE
Add clean_up_tokenization_spaces to config

### PR DESCRIPTION
**Pull request by ArthurZucker**
# What does this PR do?
DRAFT

### CODE REVIEW

**Comment by ArthurZucker** [comment]
Also linked to #20846

File: src/transformers/tokenization_utils_base.py [hunk 1]
```diff
@@ -1521,6 +1521,8 @@ def __init__(self, **kwargs):
 
         self.model_input_names = kwargs.pop("model_input_names", self.model_input_names)
 
+        # By default, cleaning tokenization spaces for both fast and slow tokenizers
+        self.clean_up_tokenization_spaces = kwargs.pop("clean_up_tokenization_spaces", True)
```
Comment (Narsil) [inline src/transformers/tokenization_utils_base.py:1525]: Will that get sent from the config ? 

I think we should add a tiny tokenizer with the config on, and confirm the thing loads a decodes properly.

Wdyt.

Then `bloom` models, `llama`  (and maybe some `gpt2`) can be updated.

Comment (ArthurZucker) [reply src/transformers/tokenization_utils_base.py:1525]: Sure! Updating in the modeling would mean a breaking changes, but if it is what we want I won't be against it

Comment (Narsil) [reply src/transformers/tokenization_utils_base.py:1525]: Wdymean the modeling ? 

I meant make sure that we can actually configure a tokenizer such that this is seemless. `tokenizer_config.json` works for me (just need to make sure it actually flows)

Comment (ArthurZucker) [reply src/transformers/tokenization_utils_base.py:1525]: Adding a test with tiny config to make sure things workout properly!

**Review by sgugger — APPROVED** [review]
Thanks for working on this! LGTM with the change in the doc suggested and the additional test suggested by Narsil.

Please make sure to fill the description before merging.

File: tests/test_tokenization_common.py [hunk 2]
```diff
@@ -3894,6 +3895,42 @@ def test_save_slow_from_fast_and_reload_fast(self):
...
+        tokenizer = BertTokenizer.from_pretrained("bert-base-uncased")
+        assert tokenizer.clean_up_tokenization_spaces == True
+        
+        tokens = tokenizer.encode("This shouldn't be! He'll go.")
+        decoded = tokenizer.decode(tokens)
+        assert decoded == "[CLS] this shouldn't be! he'll go. [SEP]"
+        
+        tokenizer.clean_up_tokenization_spaces = False
+        decoded = tokenizer.decode(tokens)
+        assert decoded ==  tokenizer.decode(tokens, clean_up_tokenization_spaces=False) == "[CLS] this shouldn ' t be ! he ' ll go . [SEP]"
```
Comment (Narsil) [inline tests/test_tokenization_common.py]: NIT: I would put those in 2 different lines for readability

File: tests/test_tokenization_common.py [hunk 3]
```diff
@@ -3894,6 +3895,42 @@ def test_save_slow_from_fast_and_reload_fast(self):
...
+        
+        with tempfile.TemporaryDirectory() as tmp_dir_2:
+            tokenizer.save_pretrained(tmp_dir_2)
+            tokenizer_fast = BertTokenizerFast.from_pretrained(tmp_dir_2)
+            del tokenizer
+    
+        decoded = tokenizer_fast.decode(tokens)
+        assert decoded == "[CLS] this shouldn ' t be ! he ' ll go . [SEP]"
```
Comment (Narsil) [inline tests/test_tokenization_common.py]: Check the value of `tokenizer_fast.clean_up_tokenization_spaces` ?

File: tests/test_tokenization_common.py [hunk 4]
```diff
@@ -3894,6 +3895,42 @@ def test_save_slow_from_fast_and_reload_fast(self):
...
+        tokenizer_fast.clean_up_tokenization_spaces = True
+        decoded = tokenizer_fast.decode(tokens)
+        assert decoded == "This shouldn't be! He'll go."
+        
+        with tempfile.TemporaryDirectory() as tmp_dir_2:
+            tokenizer_fast.save_pretrained(tmp_dir_2)
+            tokenizer = BertTokenizer.from_pretrained(tmp_dir_2)
+        
+        decoded = tokenizer.decode(tokens)
+        assert decoded == "[CLS] this shouldn ' t be ! he ' ll go . [SEP]"
```
Comment (Narsil) [inline tests/test_tokenization_common.py:3936]: Seems like a bug no ?  You just redefined `tokenizer_fast.clean_up_tokenization_spaces` as True, yet it's False after reloading.

 Maybe add a check for the value here too.

**Review by Narsil — COMMENTED** [review]
Very nice readable test !

File: tests/test_tokenization_common.py (hunk 4 above)
Comment (Narsil) [reply tests/test_tokenization_common.py:3936]: Seems code was changed, you're now setting it to false before saving (or I didn't see).

Seems good but let's check against value too I think

Comment (ArthurZucker) [reply tests/test_tokenization_common.py:3936]: Actually I do `tokenizer_fast.clean_up_tokenization_spaces = False` before saving the tokenizer_fast then from pretrained takes it!

**Review by Narsil — APPROVED** [review]
LGTM

**Pull request merged**